### ISSUE TITLE
move export map reduces to a separate github repository

### ISSUE THREAD

**Issue by cesine** [OpenSourceFieldlinguistics/FieldDB#1918]
Since we are soon going to be working more on the apps import/export functionality, this is something I would like to start thinking about strategically. In particular, Mark mentioned work he had done on this back in 2000 with Stephen Bird. I'm pretty sure the paper he was referencing was "ATLAS: A Flexible and Extensible Architecture for Linguistic Annotation" (http://arxiv.org/pdf/cs/0007022v1.pdf).

It occurred to me that we might want to initiate a separate web service software component which would have the goal of translating other data structures (e.g., LingSync's, the OLD's, ELAN's, Praat's, FLEx's, etc.) to and from this ATLAS data structure. 

-------

@jrwdunham the idea is sound and correct. the current data converters are hard to find and reuse because they are in the couchapp_dev directory. they are already webservices which run in the users databases which is most efficient in terms of indexing. but we should modify them so that they can be installed via npm and bower for maximum reuse

i made this into an issue so you will find them easier.

**Comment by cesine** [comment]
@jrwdunham  we made this this video a while back for an instructor which shows how to make custom data convertors. each time we find format we like, we add another map reduce that users can customize. we can add the atlas one too i just dont have time to help right now... i have to get the psycholing dashboard in alpha in the next 10 days.

https://www.youtube.com/watch?v=t7l_2GRumLs&index=41&list=PLUrH6CNxFDrMtraL8hTLbLsQwdw1117FT

mark probably thinks its hard to make new export routine,  and that it requires changes to the server to do it, but it doesnt. each corpus team can make their own web services, without our help or supervision (beyond showing them some tutorials for making couchdb map reduces http://www.slideshare.net/okurow/couchdb-mapreduce-13321353)

**Comment by cesine** [comment]
@cesine Great! I'll take a look at those map reduce linguistic data converters. 

However, I just want to make it clear that Mark may have no present interest in this project. He simply mentioned the ATLAS paper at the meeting since a lot of us were talking about the issue of interoperability between linguistic tools and data structures.

Also, what I was asking in particular was whether having ATLAS at the hub of a data converter would be a good idea. In the current setup you've linked to, the FieldDB data structure (such as it is) is the hub (lingua franca), correct? The use case I'm imagining is this: X is the data structure lingua franca; Y and Z can each be translated into and out of X; user U can convert his Z data to and from Y, with X as intermediary. Is this use case possible given the converter architecture you've linked to? I think it's not possible because FieldDB's data structure is too unconstrained. If I write a translator to convert my ELAN data to and from FieldDB, and my friend has done the same thing for Praat, then we are not going to be able to reliably convert ELAN to and from Praat using this service, unless we each know *how* the other has encoded certain things in the FieldDB structure, which defeats the purpose. To my understanding, solving this problem is one of the purposes of ATLAS. So it's important to determine a) whether ATLAS is well-conceived and b) whether we want to support the type of use case I have just described. If the answer to both of those questions is "yes", then we may have to re-think (re-design) the current data converter. Thoughts?

**Comment by cesine** [comment]
we'd like to have a generic convertor repo that doesn't use FieldDB as a hub. i think a converter repo
*  in javascript which is 
* bower installable 

would be handy for all the students out there who are trying to build webapps for linguistic data. it seems like every day someone tells us about find a new batch of student+prof who are making a webapp for linguistic data.

what will be the hub? I think the way you describe it (a connected graph with no obligatory hub) is the way to go about it... when the repo we created will get there, I dunno...  for now we just moved the export map reduces out into its own repo so we will build in that direction in case anyone else ever notices... 

for what its worth, did you know we already initiated a separate repo for the [Praat->JSON](https://github.com/OpenSourceFieldlinguistics/PraatTextGridJS) (its not even fielddb json, its just JSON that kind-of matches a textgrid). it has nothing to do with the FieldDB project, anyone can use it if they are using TextGrids to in a web user interface.The ELAN -> Praat -> ELAN convertors should be handled by a team who is using ELAN for annotation, if you meet some, then ask if they are doing an ELAN -> JSON convertor, or better yet, we can try to interoperate with the ELAN databasing apis that the ELAN guys have been putting significant time and thought into. 

as far as i know, i'm the only phonetics user contributing code to the app right now, so I guess its my responsibility to do all the time aligned data conversion (ie Praat and Elan lossless import and export into FieldDB) but i'm busy trying to maintain the UI code instead :( I don't use ELAN yet, but one of my informants is graduating from her Bachelors in August and iLanguage is going to fund her Post-Bac on Georgian forced alignment, so maybe that will push me to contribute more phonetics code and of course I'm going to ask her to learn ELAN since its so popular :) so progress will happen in due course...eventually...

As for re-desgn of the current data convertor:
* yep
* it should be applyable in 1 by one sequence client side
* it should be applyable in batch sequence client side or server side
* it should devide the data conversion into the [map  and reduce design pattern](http://en.wikipedia.org/wiki/MapReduce) to be able to be run on multiple nodes and clusters on large data sets
